test/converters/xcresult3/model3/conversion.go

```diff
@@ -0,0 +1,98 @@
+package model3
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+func Convert(data *TestData) (*TestSummary, error) {
+	summary := TestSummary{}
+
+	for _, testPlanNode := range data.TestNodes {
+		if testPlanNode.Type != TestNodeTypeTestPlan {
+			return nil, fmt.Errorf("test plan expected but got: %s", testPlanNode.Type)
+		}
+
+		testPlan := TestPlan{Name: testPlanNode.Name}
+
+		for _, testBundleNode := range testPlanNode.Children {
+			if testBundleNode.Type != TestNodeTypeUnitTestBundle && testBundleNode.Type != TestNodeTypeUITestBundle {
+				return nil, fmt.Errorf("test bundle expected but got: %s", testBundleNode.Type)
+			}
+
+			testBundle := TestBundle{Name: testBundleNode.Name}
+
+			for _, testSuiteNode := range testBundleNode.Children {
+				if testSuiteNode.Type != TestNodeTypeTestSuite {
+					return nil, fmt.Errorf("test suite expected but got: %s", testSuiteNode.Type)
+				}
+
+				testSuite := TestSuite{Name: testSuiteNode.Name}
+
+				for _, testCaseNode := range testSuiteNode.Children {
+					if testCaseNode.Type != TestNodeTypeTestCase {
+						return nil, fmt.Errorf("test case expected but got: %s", testCaseNode.Type)
+					}
+
+					className := strings.Split(testCaseNode.Identifier, "/")[0]
+					if className == "" {
+						// In rare cases the identifier is an empty string so we need to use the test suite name which is the
+						// same as the first part of the identifier in normal cases.
+						className = testSuiteNode.Name
+					}
+
+					testCase := TestCase{
+						Name:      testCaseNode.Name,
+						ClassName: className,
+						Time:      extractDuration(testCaseNode.Duration),
+						Result:    testCaseNode.Result,
+						Message:   extractFailureMessage(testCaseNode),
+					}
+					testSuite.TestCases = append(testSuite.TestCases, testCase)
+				}
+
+				testBundle.TestSuites = append(testBundle.TestSuites, testSuite)
+			}
+
+			testPlan.TestBundles = append(testPlan.TestBundles, testBundle)
+		}
+
+		summary.TestPlans = append(summary.TestPlans, testPlan)
+	}
+
+	return &summary, nil
+}
+
+func extractDuration(text string) time.Duration {
+	// Duration is in the format "123.456789s"
+	duration, err := time.ParseDuration(text)
+	if err != nil {
+		return 0
+	}
+
+	return duration
+}
+
+func extractFailureMessage(testNode TestNode) string {
+	childrenCount := len(testNode.Children)
+	if childrenCount == 0 {
+		return ""
+	}
+
+	lastNode := testNode.Children[childrenCount-1]
+	if lastNode.Type == TestNodeTypeRepetition {
+		return extractFailureMessage(lastNode)
+	}
+
+	failureMessage := ""
+
+	for _, child := range testNode.Children {
+		if child.Type == TestNodeTypeFailureMessage {
+			// The failure message appears in the Name field and not in the Details field.
```

Comment:
Other nodes have the useful information in the `Details` but not the `Failure Message` type which uses the `Name` field for this. Hopefully this is not bug in the tool.

Reply:
Hmm, maybe we should emit analytic logs about unexpected parse results? Like if the child.Name is empty for a FailureMessage node, or if the Details is not empty for this type?